### PR TITLE
implement `JSON_TYPE()`

### PR DESCRIPTION
**Pull request by jycor**
MySQL Docs:
https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-type

Companion PR:
https://github.com/dolthub/vitess/pull/314

### CODE REVIEW

File: sql/expression/function/json/json_type_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	_, err := NewJSONType()
+	require.True(t, errors.Is(err, sql.ErrInvalidArgumentNumber))
+
+	f1 := buildGetFieldExpressions(t, NewJSONType, 1)
+	testCases := []struct {
+		f   sql.Expression
+		row sql.Row
+		exp interface{}
+		err bool
+	}{
```
Comment (nicktobey) [inline sql/expression/function/json/json_type_test.go:38]: Add a test here for DECIMAL. You may have to explicitly pass in a Decimal object to the Row.

**Review by nicktobey — APPROVED** [review]

File: sql/expression/function/json/json_type_test.go (hunk 1 above)
Comment (jycor) [reply sql/expression/function/json/json_type_test.go:38]: Raw decimals can't be used in `json_type()`; it has to go through a `cast(<dec> to json)`.
There is an enginetest for this

File: enginetest/queries/queries.go [hunk 2]
```diff
@@ -9354,6 +9354,84 @@ from typestable`,
...
+	{
+		Query: "SELECT json_type(cast(cast('2001-01-01 12:34:56.123456' as date) as json));",
+		Expected: []sql.Row{
+			{"DATE"},
+		},
+	},
+	{
+		Query: "select json_type(cast(cast(1 as unsigned) as json));",
+		Expected: []sql.Row{
+			{"UNSIGNED INTEGER"},
```
Comment (fulghum) [inline enginetest/queries/queries.go:9414]: oooh, nice work! The docs don't mention "UNSIGNED INTEGER" as a possible response, but that's exactly what MySQL returns! Good job reversing engineering MySQL's behavior here.

File: sql/expression/function/json/json_type.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+				return "TIME", nil
+			}
+		}
+		return "STRING", nil
+	case []interface{}:
+		return "ARRAY", nil
+	case map[string]interface{}:
+		return "OBJECT", nil
+	default:
+		return "OPAQUE", nil
```
Comment (fulghum) [inline sql/expression/function/json/json_type.go:130]: Any way to trigger `OPAQUE`? I don't think I saw that in any of the tests.

File: enginetest/queries/queries.go [hunk 4]
```diff
@@ -9559,6 +9637,34 @@ FROM mytable;`,
...
 		Expected: []sql.Row{{"myview1", "YES"}, {"myview2", "YES"}, {"myview3", "NO"}, {"myview4", "NO"}, {"myview5", "YES"}},
 	},
+	// time to json cast is broken
+	{
+		Query: "SELECT json_type(cast(cast('2001-01-01 12:34:56.123456' as time) as json));",
+		Expected: []sql.Row{
+			{"TIME"},
+		},
+	},
+	// missing year cast
```
Comment (fulghum) [inline enginetest/queries/queries.go:9647]: Are we still missing something for this? I saw we added cast to year syntax in vitess and some code in GMS for casting. Just curious what else is needed for this one.

File: sql/expression/function/json/json_type.go (hunk 3 above)
Comment (jycor) [reply sql/expression/function/json/json_type.go:130]: I haven't found a query that spits out `OPAQUE` in MySQL or dolt

File: enginetest/queries/queries.go [hunk 5]
```diff
@@ -9354,6 +9354,72 @@ from typestable`,
 			{false},
 		},
 	},
+	{
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply enginetest/queries/queries.go:9357]: I think I was wrong about being able to use scientific notation here.

I think we should add the following tests, to make sure that we respect round-trip conversions of types:

`SELECT json_type(json_extract(json_object("a", cast(10 as double)), "$.a"))` -> "DOUBLE"
`SELECT json_type(json_extract(json_object("a", cast(10 as unsigned)), "$.a"))` -> "UNSIGNED INTEGER"
`SELECT json_type(json_extract(json_object("a", cast(10 as signed)), "$.a"))` -> "INTEGER"
`SELECT json_type(json_extract(json_object("a", cast(10 as decimal)), "$.a"))` -> "DECIMAL"

It's possible that the way we currently store JSON causes some of these to not work properly. We can disable those tests if it's out of scope to fix them now. But we should still have them as tests.

File: enginetest/queries/queries.go (hunk 4 above)
Comment (jycor) [reply enginetest/queries/queries.go:9647]: oops this test should not be skipped; will fix

File: sql/expression/function/json/json_type_test.go (hunk 1 above)
Comment (nicktobey) [reply sql/expression/function/json/json_type_test.go:38]: Ah, gotcha. Sounds good then.

File: enginetest/queries/queries.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jycor) [reply enginetest/queries/queries.go:9357]: didn't see this comment sorry!

test are added here:
https://github.com/dolthub/go-mysql-server/pull/2355

**Pull request merged**